### PR TITLE
feat: Add useReducer

### PR DESCRIPTION
**Pull request by Donskelle**
This Hook is inspired by Reacts buildin useReducer Hook. Its an easy way to centralize logic and use it in sub components without needing reference to the function.

The Name Reducer is a little misleading because its not a pure function to update state. Its mutating the reactive vue variable.

Maybe the name useActionState would be better? What do you think about this in general?

I had some issue with ts

### CODE REVIEW

**Comment by antfu** [comment]
Thanks for the PR!

I am personally not a big fan of reducers as they make the types harder. But I am happy to have this included, as long as we can have a proper way to either auto infer the types or let users able to define the types.

I am fine with both `useReducer` and `useActionState`. If we go with `useReducer`, please add a line in docs addressing the difference.

BTW, can you run the linter to fix the style error? Thanks :)

**Comment by Donskelle** [comment]
Ok, ya linting is no problem.
TS I need some days to research

File: packages/core/useReducer/index.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { reactive, UnwrapRef } from 'vue-demi'
+
+export interface Action {
+  type: string
+  [propName: string]: any
+}
```
Comment (Donskelle) [inline packages/core/useReducer/index.ts:6]: Is this okay? Or do you think it should be more flexible?

File: packages/core/useReducer/ComplexDemo.tsx [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    data: {
+      type: Array,
+      default: () => [],
+    },
+  },
+  render(this: Vue) {
+    return (
+      <ul>
+        {this.$props.data.map((ele: number) => (
+          <ListItem id={ele} />
```
Comment (Donskelle) [inline packages/core/useReducer/ComplexDemo.tsx:73]: `##[error]packages/core/useReducer/ComplexDemo.tsx(73,21): error TS2769: No overload matches this call.`
I dont understand how i can solve this.

File: packages/core/useReducer/index.ts (hunk 1 above)
Comment (antfu) [reply packages/core/useReducer/index.ts:6]: Just checked react's approach https://github.com/DefinitelyTyped/DefinitelyTyped/blob/bbc8c7c7b7e92ba094ea349d56977e00f6f7f42d/types/react/index.d.ts#L927-L1006
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/bbc8c7c7b7e92ba094ea349d56977e00f6f7f42d/types/react/index.d.ts#L871-L880

Which I think you can refer to :)

Comment (Donskelle) [reply packages/core/useReducer/index.ts:6]: Sorry, wasnt very motivated to do this in the last month and still havent figured out how i wanna do the reset. I'll close this for now, sorry

Comment (Donskelle) [reply packages/core/useReducer/index.ts:6]: Enjoy holiday and thanks for your awesome open source work !

**Pull request closed without merging**